python/ray/experimental/workflow/workflow_storage.py

```diff
@@ -0,0 +1,268 @@
+
+    def write_step_output(self, step_id: StepID, output: Any) -> None:
+        """Save the output of a workflow step.
+
+        Args:
+            output: The output object.
+        """
+        # convert
+        obj = ray.get(output) if isinstance(output, ray.ObjectRef) else output
+        step_dir = self._steps_dir / step_id
+        self._storage.write_object_atomic(obj, step_dir / STEP_OUTPUT)
+
+    def read_step_output(self, step_id: StepID) -> Any:
+        """Get the output of the workflow step from checkpoint.
+
+        Args:
+            step_id: ID of the workflow step.
+
+        Returns:
+            Output of the workflow step.
+        """
+        return self._storage.read_object(
+            self._steps_dir / step_id / STEP_OUTPUT)
+
+    def read_step_func_body(self, step_id: StepID) -> Callable:
+        """Get the function body of the workflow step.
+
+        Args:
+            step_id: ID of the workflow step.
+
+        Returns:
+            A callable function.
+        """
+        return self._storage.read_object(
+            self._steps_dir / step_id / STEP_FUNC_BODY)
+
+    def read_step_args(self, step_id: StepID) -> Tuple[List, Dict[str, Any]]:
+        """Get the input arguments of the workflow step. This must be
+        done under a serialization context, otherwise the arguments would
+        not be reconstructed successfully.
+
+        Args:
+            step_id: ID of the workflow step.
+
+        Returns:
+            Args and kwargs.
+        """
+        return self._storage.read_object(self._steps_dir / step_id / STEP_ARGS)
+
+    def read_object_ref(self, object_id) -> ray.ObjectRef:
+        """Get the input object ref.
+
+        Returns:
+            The object ref.
+        """
+        obj = self._storage.read_object(self._objects_dir / object_id)
+        return ray.put(obj)  # simulate an ObjectRef
+
+    def get_entrypoint_step_id(self) -> StepID:
+        step_id = self._locate_output_step(self._steps_dir)
+        if not isinstance(step_id, str):
+            raise ValueError("Cannot locate workflow entrypoint step.")
+        return step_id
+
+    def _locate_output_step(self, step_dir: pathlib.Path) -> Optional[StepID]:
+        """Locate where the output comes from with the given step dir.
+
+        Args:
+            step_dir: The path to search answers.
+
+        Returns
+            If found, return the ID of the step that produce results.
+            Otherwise return None.
+        """
+        # read outputs forward file (take a shortcut)
+        try:
+            metadata = self._storage.read_json(step_dir / STEP_OUTPUTS_FORWARD)
+            return metadata["step_id"]
+        except (FileNotFoundError, json.JSONDecodeError):
+            pass
+        # read outputs metadata
+        try:
+            metadata = self._storage.read_json(
+                step_dir / STEP_OUTPUTS_METADATA)
+            return metadata["step_id"]
```

Comment:
It's confusing to have two different output files. Could we just have a single one, and call it `output_step_id`?